cylc/flow/graph_parser.py

```diff
@@ -458,68 +572,150 @@ def _families_all_to_all(self, expr, rights, info, family_trig_map):
                 m_expr = []
                 for mem in self.family_map[name]:
                     m_info.append((mem, offset, ttype))
-                    m_expr.append("%s%s%s" % (mem, offset, ttype))
-                this = r'\b%s%s%s\b' % (name, re.escape(offset), trig)
-                if extn == self.__class__.FAM_TRIG_EXT_ALL:
+                    m_expr.append(f"{mem}{offset}:{ttype}")
+                this = r'\b%s%s:%s\b' % (
+                    name,
+                    re.escape(offset),
+                    trig
+                )
+                if mem_all:
                     that = '(%s)' % '&'.join(m_expr)
-                elif extn == self.__class__.FAM_TRIG_EXT_ANY:
+                else:
                     that = '(%s)' % '|'.join(m_expr)
                 n_expr = re.sub(this, that, n_expr)
                 n_info += m_info
             else:
                 n_info += [(name, offset, trig)]
-        self._add_trigger(expr, rights, n_expr, n_info)
 
-    def _add_trigger(self, orig_expr, rights, expr, info):
+        self._compute_triggers(expr, rights, n_expr, n_info)
+
+    def _set_triggers(
+        self, output_map, name, output, optional, suicide, trigs, expr,
+        orig_expr, family=False
+    ):
+        """Record parsed triggers and outputs."""
+
+        # Check suicide triggers
+        with contextlib.suppress(KeyError):
+            osuicide = self.triggers[name][expr][1]
+            # This trigger already exists, so we must have both
+            # "expr => member" and "expr => !member" in the graph,
+            # or simply a duplicate trigger not recognized earlier
+            # because of parameter offsets.
+            if not expr:
+                pass
+            elif suicide is not osuicide:
+                oexp = re.sub(r'(&|\|)', r' \1 ', orig_expr)
+                oexp = re.sub(r':succeeded', '', oexp)
+                raise GraphParseError(
+                    f"{oexp} can't trigger both {name} and !{name}"
+                )
+
+        # Record triggers
+        self.triggers.setdefault(name, {})
+        self.triggers[name][expr] = (trigs, suicide)
+        self.original.setdefault(name, {})
+        self.original[name][expr] = orig_expr
+
+        if family and output == "" or suicide:
+            # Do not infer output optionality from suicide triggers
+            # or from a family name on the right side.
+            return
+
+        # Add or check {(name, output): optional} in output_map.
+        try:
+            already = output_map[(name, output)]
+        except KeyError:
+            # Not already in map; add it.
+            output_map[(name, output)] = optional
+        else:
+            # Already in map; check consistency with previous value,
+            if already != optional:
+                if not family:
+                    raise GraphParseError(
+                        f"{name}:{output} can't be both optional and required"
+                    )
```

Comment:
Codecov is complaining about these - are there not unit tests for these exceptions? (The coverage has been known to misbehave but it should have been fixed by #4348

Reply:
Oops, more unit tests to come...